### ISSUE TITLE
Unable to set CORS headers

### ISSUE THREAD

**Issue by tedivm** [tiangolo/fastapi#133]
**Describe the bug**

Attempting to add CORS headers does not work.

**To Reproduce**
There are two ways to add the CORS headers- using the Starlette CORS middleware, or by manually adding headers. Neither work.

```python
from fastapi import FastAPI
from starlette.middleware.cors import CORSMiddleware

app = FastAPI()
app.add_middleware(CORSMiddleware, allow_origins=['*'])

@app.get('/ping', content_type=UJSONResponse)
def health_check():
    return UJSONResponse({'status': 'ok'}, headers={'Access-Control-Allow-Origin': '*'})

```

It looks like the middleware is just completely ignored, and the headers defined in the routes don't work when the "OPTIONS" verb is used.

**Expected behavior**
The CORS headers should be set.

**Environment:**

Tested with `tiangolo/uvicorn-gunicorn-fastapi:python3.6`.

**Comment by tedivm** [comment]
My mistake- I was trying to POST but hadn't actually set that verb to allowed.

**Comment by tiangolo** [comment]
Thanks for reporting back that you were able to fix it.

Also, you can check the project generator section that handles that: https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/main.py#L20-L26

**Comment by tedivm** [comment]
@tiangolo - that's literally how I found my mistake! Having that starter app really helped out.

**Comment by tiangolo** [comment]
Aha! Hehe great.

**Comment by Mbugua** [comment]
A neat fix is also to ensure you have imported the response in the respective  endpoint  i.e
`from starlette.responses import JSONResponse, Response, UJSONResponse`

**Comment by leblancfg** [comment]
If you're reading this, here's the minimal example I had to set up to get CORS working on my end. Reading the docs, it seemed implied that the OPTIONS route was self-generated. It is not, and it is needed when [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) is used across origins.

~~~python
from fastapi import FastAPI, File, UploadFile
from starlette.responses import UJSONResponse
from starlette.middleware.cors import CORSMiddleware

app = FastAPI()
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],  # FIXME for production
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


@app.options('/test')
def preflight():
    return UJSONResponse({'status': 'ok'}, headers={'Access-Control-Allow-Origin': '*'})


@app.post("/test")
def test_function(file: UploadFile = File(...),):
    # Do work
    test_item = do_work(file.file)
    return test_item

~~~

**Comment by CoderCharm** [comment]
I have the same problem。

## Environment
```
fastapi==0.55.1
python3.7
```
my code 
```
app.add_middleware(
    CORSMiddleware,
    allow_origins=['*'],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```
`GET`requests can cross domains, but `POST` requests cannot。


I saw that there was a `allow_origin_regex ` way in the document, so I did it and solved my problem, `GET` and `POST` can be cross domain.

```
app.add_middleware(
    CORSMiddleware,
    # allow_origins=origins,
    allow_origin_regex='https?://.*',
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```

I don't know why `allow_origins=['*'] 'will be invalid.

**Comment by ccharlesgb** [comment]
Hi CoderCharm, we ran into this issue a while back it is due to CORS policies: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSNotSupportingCredentials

If the request sends credentials then you are not allowed any origin (i.e `allow_origins=['*']`) and the origin must be 'reflected' back by the Middleware dynamically (Which is what the regex does): 

https://github.com/encode/starlette/blob/93878323e57e0bab92b4622849c67f5a7c96b24e/starlette/middleware/cors.py#L107

**Comment by CoderCharm** [comment]
@ccharlesgb Thank you for answering my doubts.

**Comment by synchronizing** [comment]
tl;dr of @ccharlesgb answer.

```python
app.add_middleware(CORSMiddleware, allow_all_origins=True)
```
